### DISCUSSION TITLE
Is there a corref model updated with Spacy 3.7?

### DISCUSSION THREAD

**Discussion by lingvisa** [explosion/spaCy#13572] [Language Support]
I am using Spacy 3.7.5, and I want to use the coref cabability, and following this instruction:
https://github.com/explosion/spaCy/discussions/11585

I had this:
```
nlp = spacy.load("en_core_web_trf")

#assert "transformer" not in nlp.pipe_names
nlp_coref = spacy.load("en_coreference_web_trf")
#nlp.add_pipe("transformer", source=nlp_coref)
nlp.add_pipe("coref", source=nlp_coref)
nlp.add_pipe("span_resolver", source=nlp_coref)
nlp.add_pipe("span_cleaner", source=nlp_coref)
```

To install the coref dependency:
` pip install https://github.com/explosion/spacy-experimental/releases/download/v0.6.1/en_coreference_web_trf-3.4.0a2-py3-none-any.whl`

This will downgrade my spacy from 3.75 to 3.44. Is there an updated en_coreference_web_trf that I can use with Spacy 3.7 above?

**Comment by davidress-ILW** [comment]
@lingvisa 

See:
[https://github.com/explosion/spaCy/issues/13111](url)

There is a note on using spacy 3.7 with the en_coreference_web_trf model that was compiled with spacy 3.4.

**Reply by lingvisa** [reply]
@davidress-ILW That link is dead. Is there an updated one?

**Reply by lingvisa** [reply]
Also, it seems there are 3 versions of coref:
1) spacy implementation
2) neuralcoref
3) https://spacy.io/universe/project/coreferee

I understand 1) replaces 2. What about the 3rd option "coreferee"?

**Reply by davidress-ILW** [reply]
@lingvisa Hmm,, odd that the link is dead. If you want to manually find it, go to the Issues page, search the closed issues and for the subject "coref" for the title "Pre-trained coreference pipeline incompatible with spaCy > 3.4 #13111" 

I am using en_coreference_web_trf from spacy/explosion with spacy 3.7.5 at the moment. My installation steps are:

1. python -m pip install https://github.com/explosion/spacy-experimental/releases/download/v0.6.1/en_coreference_web_trf-3.4.0a2-py3-none-any.whl
2.  python -m pip install spacy==3.7.5 -U
3. python -m spacy download en_core_web_trf

Then I am using Python 3.10 and the following code to link it all together:

        nlp = spacy.load("en_core_web_trf")
        nlp_coref = spacy.load("en_coreference_web_trf")

        # use replace_listeners for the coref components
        nlp_coref.replace_listeners("transformer", "coref", ["model.tok2vec"])
        nlp_coref.replace_listeners("transformer", "span_resolver", ["model.tok2vec"])

        # we won't copy over the span cleaner - this keeps the head cluster information, which we want
        nlp.add_pipe("merge_entities")
        nlp.add_pipe("coref", source=nlp_coref)
        nlp.add_pipe("span_resolver", source=nlp_coref)


In running the above code, there are warnings that a model that was trained with spacy 3.4 may not be 100% compatible with spacy 3.7.5 are displayed, but that is something I must be comfortable living with as I need coreference resolution.  

I am using en_coreference_web_trf over coreferee due to some coreference concerns I was seeing during testing between the two with spacy 3.7.5. Here is an example concern: 

- Sentence: To finance the company's creation, Jobs sold his Volkswagen Bus, and Wozniak sold his HP-65 calculator.
- Coreferee output: To finance the company 's creation, Jobs sold Jobs Volkswagen Bus, and Wozniak sold **Jobs** HP-65 calculator.
- en_coreference_web_trf output: To finance the company's creation, Jobs sold Jobs Volkswagen Bus, and Wozniak sold **Wozniak** HP-65 calculator.

In the above example, coreferee did not correctly corefence the owner of the HP-65 calculator. This is just one of many that caused me to start using en_coreference_web_trf over coreferee.

Aside from my testing preference, what I have learned about coreference using spacy is as follows:

- coreferee is no longer being maintained.
- en_coreference_web_trf which is a part of spacy/explosion, will eventually be rolled into spacy proper (see #13111)

**Comment by lingvisa** [comment]
@davidress-ILW Your code above works well. I have two questions on how they work internally:
1) The replace_listener configures coref to listen to the transformer in nlp_coref, and after coref is added to 'nlp', coref will use transformer from 'nlp', not 'nlp_coref', then what's the purpose of the listener? 
2) coref_cluster and coref_head_cluster? Which should I use in my code?
